File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/JavaClassProxy.java
package com.github.ol_loginov.heaplibweb.services.proxies;

import com.github.ol_loginov.heaplibweb.repository.heap.HeapScope;
import com.github.ol_loginov.heaplibweb.repository.heap.JavaClassEntity;
import lombok.RequiredArgsConstructor;
import org.netbeans.lib.profiler.heap.Field;
import org.netbeans.lib.profiler.heap.FieldValue;
import org.netbeans.lib.profiler.heap.Instance;
import org.netbeans.lib.profiler.heap.JavaClass;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;

@RequiredArgsConstructor
public class JavaClassProxy implements JavaClass {
	private final JavaClassEntity entity;
	private final HeapScope scope;

	public static JavaClass wrap(JavaClassEntity other, HeapScope scope) {
		return new JavaClassProxy(other, scope);
	}

	private JavaClass wrap(JavaClassEntity other) {
		return wrap(other, scope);
	}

	@Override
	public long getAllInstancesSize() {
		return entity.getAllInstancesSize();
	}

	@Override
	public boolean isArray() {
		return entity.isArray();
	}

	@Override
	public String getName() {
		return entity.getName();
	}

	@Override
	public int getInstanceSize() {
		return entity.getInstanceSize();
	}

	@Override
	public int getInstancesCount() {
		return entity.getInstancesCount();
	}

	@Override
	public long getRetainedSizeByClass() {
		return HeapScope.shouldBeReady(entity.getRetainedSizeByClass());
	}

	@Override
	public long getJavaClassId() {
		return entity.getJavaClassId();
	}

	@Override
	public JavaClass getSuperClass() {
		if (entity.getSuperClassId() == null) {
			return null;
		}
		return scope.getJavaClasses().findById(new JavaClassEntity.PK(entity.getHeapId(), entity.getSuperClassId()))
			.map(this::wrap)
			.orElse(null);
	}

	@Override
	public Collection<JavaClass> getSubClasses() {
		return scope.getJavaClasses().streamAllByHeapIdAndSuperClassId(entity.getHeapId(), entity.getJavaClassId())
			.map(this::wrap)
			.toList();
	}

	@Override
	public List<Field> getFields() {
		return scope.getFields().streamAllByHeapIdAndDeclaringClassId(entity.getHeapId(), entity.getJavaClassId())
			.map(e -> (Field) new FieldProxy(e, scope))
			.toList();
	}

	@Override
	public List<Instance> getInstances() {
		return scope.getInstances().streamAllByHeapIdAndJavaClassId(entity.getHeapId(), entity.getJavaClassId())
			.map(e -> InstanceProxy.wrap(e, scope))
			.toList();
	}

	@Override
	public List<FieldValue> getStaticFieldValues() {
		return scope.getFieldValues().streamStaticFieldValues(entity.getJavaClassId())
			.map(e -> FieldValueProxy.wrap(entity.getHeapId(), e, scope))
			.toList();
	}

	@Override
	public Object getValueOfStaticField(String name) {
		return scope.getFieldValues().findStaticByClassAndFieldName(entity.getJavaClassId(), name)
			.map(e -> FieldValueProxy.getValueObject(e, entity.getHeapId(), scope))
			.orElse(null);
	}


	@Override
	public Instance getClassLoader() {
		return (Instance) getValueOfStaticField("<classLoader>");
	}

	@Override
	public Iterator<Instance> getInstancesIterator() {
		throw new UnsupportedOperationException();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/JavaClassProxy.kt
package com.github.ol_loginov.heaplibweb.services.proxies

import com.github.ol_loginov.heaplibweb.repository.heap.ClassEntity
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories
import com.github.ol_loginov.heaplibweb.services.ValueNotReadyException
import com.github.ol_loginov.heaplibweb.services.proxies.FieldValueProxy.Companion.getValueObject
import org.netbeans.lib.profiler.heap.Field
import org.netbeans.lib.profiler.heap.FieldValue
import org.netbeans.lib.profiler.heap.Instance
import org.netbeans.lib.profiler.heap.JavaClass

class JavaClassProxy(
    private val entity: ClassEntity,
    private val heapRepositories: HeapRepositories
) : JavaClass {

    companion object {
        @JvmStatic
        fun wrap(other: ClassEntity, heapRepositories: HeapRepositories): JavaClass = JavaClassProxy(other, heapRepositories)
    }

    private fun wrap(other: ClassEntity): JavaClass = wrap(other, heapRepositories)

    override fun getAllInstancesSize(): Long = entity.allInstancesSize ?: throw ValueNotReadyException()
    override fun isArray(): Boolean = entity.array
    override fun getName(): String = entity.name
    override fun getInstanceSize() = entity.instanceSize
    override fun getInstancesCount(): Int = entity.instancesCount
    override fun getRetainedSizeByClass(): Long = entity.retainedSizeByClass ?: throw ValueNotReadyException()
    override fun getJavaClassId(): Long = entity.id
    override fun getSuperClass(): JavaClass? {
        return entity.superClassId
            ?.let { heapRepositories.classes.findByIdSure(it) }
            ?.let { wrap(it) }
    }

    override fun getSubClasses(): Collection<JavaClass> = heapRepositories
        .classes.streamAllBySuperClassId(entity.id)
        .map(this::wrap)
        .toList()

    override fun getFields(): List<Field> = heapRepositories
        .fields.findAllByDeclaringClassIdOrderById(entity.id)
        .map { FieldProxy(it, heapRepositories) }

    override fun getInstances(): List<Instance> = heapRepositories
        .instances.streamAllByJavaClassId(entity.id)
        .map { InstanceProxy.wrap(it, heapRepositories) }
        .toList()


    override fun getStaticFieldValues(): List<FieldValue> {
        return heapRepositories.fieldValues.streamStaticFieldValues(entity.id)
            .map { FieldValueProxy.wrap(it, heapRepositories) }
            .toList()
    }

    override fun getValueOfStaticField(name: String): Any? = heapRepositories
        .fieldValues.findStaticByClassAndFieldName(entity.id, name)
        ?.let { getValueObject(it, heapRepositories) }

    override fun getClassLoader(): Instance? = getValueOfStaticField("<classLoader>") as Instance?
    override fun getInstancesIterator(): Iterator<Instance> = throw UnsupportedOperationException()
}
